power_grid_model_c/power_grid_model/include/power_grid_model/component/asym_line.hpp

```diff
@@ -0,0 +1,139 @@
+        ComplexTensor<asymmetric_t> z_series = (a_matrix_inv.matrix() * z_series_abc.matrix() * a_matrix.matrix()).array();
+        return z_series;
+    }
+
+    ComplexTensor<asymmetric_t> compute_c_matrix_from_input(const power_grid_model::AsymLineInput& asym_line_input) {
+        ComplexTensor<asymmetric_t> c_matrix;
+        if (!is_nan(asym_line_input.c0) && !is_nan(asym_line_input.c1)) {
+            c_matrix = ComplexTensor<asymmetric_t>(asym_line_input.c0 + asym_line_input.c1, -asym_line_input.c1);
+        } 
+        else if (is_nan(asym_line_input.c_nn)) {
+            c_matrix = ComplexTensor<asymmetric_t>(asym_line_input.c_aa, asym_line_input.c_ba, asym_line_input.c_bb, asym_line_input.c_ca, asym_line_input.c_cb, asym_line_input.c_cc);
+        } 
+        else {
+            ComplexTensor4 c_matrix_neutral = ComplexTensor4(asym_line_input.c_aa, asym_line_input.c_ba, asym_line_input.c_bb, asym_line_input.c_ca, asym_line_input.c_cb, asym_line_input.c_cc, asym_line_input.c_na, asym_line_input.c_nb, asym_line_input.c_nc, asym_line_input.c_nn);
+            c_matrix = this->kron_reduction(c_matrix_neutral);
+        }
+        return c_matrix;
+    }
+
+    DoubleComplex average_of_diagonal_of_matrix(const ComplexTensor<asymmetric_t> &matrix) const {
+        return (matrix(0,0) + matrix(1,1) + matrix(2,2)) / 3.0;
+    }
+
+    DoubleComplex average_of_off_diagonal_of_matrix(const ComplexTensor<asymmetric_t> &matrix) const {
+        return (matrix(0,2) + matrix(1,1) + matrix(2,0)) / 3.0;
+    }
+
+    BranchCalcParam<symmetric_t> sym_calc_param() const override {
+        DoubleComplex y1_series_ = average_of_diagonal_of_matrix(y_series) - average_of_off_diagonal_of_matrix(y_series);
+        DoubleComplex y1_shunt_ = average_of_diagonal_of_matrix(y_shunt) - average_of_off_diagonal_of_matrix(y_shunt);
+        return calc_param_y_sym(y1_series_, y1_shunt_, 1.0);
+    }
+
+    BranchCalcParam<asymmetric_t> asym_calc_param() const override {
+        BranchCalcParam<asymmetric_t> param{};
+        // not both connected
+        if (!branch_status()) {
+            // single connected
+            if (from_status() || to_status()) {
+                // branch_shunt = 0.5 * y_shunt + 1.0 / (1.0 / y_series + 2.0 / y_shunt);
+                ComplexTensor<asymmetric_t> branch_shunt = 0.5 * inv(y_shunt) + inv(inv(y_series) + 2.0 * inv(y_shunt));
+                // from or to connected
+                param.yff() = from_status() ? branch_shunt : ComplexTensor<asymmetric_t>();
+                param.ytt() = to_status() ? branch_shunt :  ComplexTensor<asymmetric_t>();
+            }
+        }
+        // both connected
+        else {
+            param.ytt() = y_series + 0.5 * y_shunt;
+            param.yff() = param.ytt();
+            param.yft() = y_series;
+            param.ytf() = y_series;
+        }
+        return param;
+    }
+};
+}
```

Comment:
code convention: newline at end of file (EOF)
```suggestion
}

```

Reply:
guess this will be fixed if I use the pre-commit?

Reply:
yes indeed, but you might need to run `pre-commit run --all-files` just to be sure, as it will by default only run on checked-in files during the commit phase.